runtime/interpreter/interpreter.go

```diff
@@ -978,6 +980,46 @@ func (interpreter *Interpreter) declareAttachmentValue(
 	return interpreter.declareCompositeValue(declaration, lexicalScope)
 }
 
+// evaluateDefaultDestroyEvent evaluates all the implicit default arguments to the default destroy event.
+//
+// the handling of default arguments makes a number of assumptions to simplify the implementation;
+// namely that a) all default arguments are lazily evaluated at the site of the invocation,
+// b) that either all the parameters or none of the parameters of a function have default arguments,
+// and c) functions cannot currently be explicitly invoked if they have default arguments
+//
+// if we plan to generalize this further, we will need to relax those assumptions
+func (declarationInterpreter *Interpreter) evaluateDefaultDestroyEvent(
+	containingResourceComposite *CompositeValue,
+	eventDecl *ast.CompositeDeclaration,
+	declarationActivation *VariableActivation,
+) (arguments []Value) {
+	parameters := eventDecl.DeclarationMembers().Initializers()[0].FunctionDeclaration.ParameterList.Parameters
+
+	declarationInterpreter.activations.Push(declarationActivation)
+	defer declarationInterpreter.activations.Pop()
+
+	var self MemberAccessibleValue = containingResourceComposite
+	if containingResourceComposite.Kind == common.CompositeKindAttachment {
+		var base *EphemeralReferenceValue
+		base, self = attachmentBaseAndSelfValues(declarationInterpreter, containingResourceComposite)
+		declarationInterpreter.declareVariable(sema.BaseIdentifier, base)
+	}
+	declarationInterpreter.declareVariable(sema.SelfIdentifier, self)
```

Comment:
I'm not sure, but it might be necessary to push new activation _on top_ of the declaration activation, i.e. have a separate activation for the scope of just the event constructor arguments. The activation that is pushed onto the activation stack is currently the declaration activation, and then it is mutated below and `self` (and `base` for attachments) are declared in it – that should happen in a fresh activation.

I think, but am not sure and don't know if it's an issue, that the current implementation basically "leaks" `self` (and `base` for attachments) to the code following the `ResourceDestroyed` event declaration. I can't think of a test case of the top of my head which could test it though.

Maybe also double check this is done for the checker part of this feature.